### PR TITLE
Fix retain_grad() documentation for leaf vs non-leaf tensors

### LINKED ISSUE #3652: retain_grad() documentation misleading about non-leaf tensors
The documentation says:

"If you call retain_grad() on a non-leaf node, it results in a no-op."

This is misleading or incomplete.
retain_grad() on a non-leaf tensor with requires_grad=True does not result in a no-op — it correctly retains gradients.

Only calling retain_grad() on:

a leaf tensor is a no-op

a tensor with requires_grad=False throws an error

### PR DESCRIPTION
**Pull request by patrocinio**
The documentation incorrectly stated that calling retain_grad() on a non-leaf node results in a no-op. This is misleading because:

- retain_grad() on a non-leaf tensor with requires_grad=True correctly retains gradients (not a no-op)
- retain_grad() on a leaf tensor is a no-op (already retains by default)
- retain_grad() on a tensor with requires_grad=False throws an error

Updated the documentation and summary table to accurately reflect this behavior.

Fixes #3652

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

cc @jbschlosser

### CODE REVIEW

**Comment by svekars** [comment]
cc: @j-silv

**Review by soulitzer — APPROVED** [review]
Thanks!
